### PR TITLE
refactor: add gsl::not_null to get compile time / run time pointer guarantees

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented
Current implementation relies either on asserts or sometimes checks then returning a special value; In the case of asserts (or no assert where we use the value without checks) it'd be better to make it explicit to function caller that the ptr must be not_null; otherwise gsl::not_null will call terminate. 

See https://github.com/microsoft/GSL/blob/main/docs/headers.md#user-content-H-pointers-not_null and https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-nullptr

I'm interested in a conceptual review; specifically on if this is beneficial over just converting these ptrs to be a reference?

## What was done?
 *Partial* implementation on using gsl::not_null in dash code

## How Has This Been Tested?
Building

## Breaking Changes
None

## Checklist:
  _Go over all the following points, and put an `x` in all the boxes that apply._
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Comment by knst** [comment]
>  otherwise gsl::not_null will call terminate.

I think `assert` is better than std::terminate, because with assert you can see backtrace when app is running in debugger.

**Comment by PastaPastaPasta** [comment]
> > otherwise gsl::not_null will call terminate.
> 
> I think `assert` is better than std::terminate, because with assert you can see backtrace when app is running in debugger.

Well, we could change [this](https://github.com/dashpay/dash/pull/5595/commits/74654d4489b08a3c0cf9b2d19f818592737ae253#diff-da98b238347b6ff99b605135e83842d5d2ca7e29b3a612be14d59989d7980886R116) to just be assert(false) then instead of calling std::terminate

**Comment by knst** [comment]
> Well, we could change [this](https://github.com/dashpay/dash/pull/5595/commits/74654d4489b08a3c0cf9b2d19f818592737ae253#diff-da98b238347b6ff99b605135e83842d5d2ca7e29b3a612be14d59989d7980886R116) to just be assert(false) then instead of calling std::terminate

Will we have in this case any error message in console (without debugger) such as "pointer X at file.cpp:NNN is null" also same as with assert? If so it's fine then

**Comment by PastaPastaPasta** [comment]
Please take a look at this further set of changes and LMK what you think? it uses an early implementation of c++20's std::source_location that I think should be low cost; provides output now like
`ERROR: error detected null not_null detected at dsnotificationinterface.cpp:41:57:AcceptedBlockHeader`

File: src/gsl/assert.h [hunk 1]
```diff
@@ -108,23 +110,25 @@ namespace gsl
 
 #endif // defined(GSL_MSVC_USE_STL_NOEXCEPTION_WORKAROUND)
 
-        [[noreturn]] inline void terminate() noexcept
+        [[noreturn]] inline void terminate(nostd::source_location loc) noexcept
         {
 #if defined(GSL_MSVC_USE_STL_NOEXCEPTION_WORKAROUND)
             (*gsl::details::get_terminate_handler())();
 #else
+            std::cout << "ERROR: error detected null not_null detected at " << loc.file_name() << ":" << loc.line() << ":"
```
Comment (knst) [inline src/gsl/assert.h:118]: Std::cerr I think should be, also std::endl to flush buffer

**Review by knst — COMMENTED** [review]
Concept ACK if it works as described

Btw, looks like magic for location of call

**Comment by knst** [comment]
> I think assert is better than std::terminate, because with assert you can see backtrace when app is running in debugger.

As I checked, at least `gdb` catches `std::terminate` without issues:
```
$ gdb -q main
Reading symbols from main...
(gdb) r
Starting program: /tmp/main 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
terminate called without an active exception

Program received signal SIGABRT, Aborted.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:44
44      ./nptl/pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=<optimized out>, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff783c406 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff782287c in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff7ca4ee6 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff7cb6e9c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x00007ffff7cb6f07 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#8  0x0000555555555156 in main () at main.cpp:4
(gdb) quit
A debugging session is active.
```

main.cpp:
```
#include <exception>

int main() {
    std::terminate();
}
```
So, std::terminate(), abort() and assert() behave are very similar in gdb and std::terminate seems completely fine.

_Probably I got confused by `exit()` that is used in libjpeg/jpegturbo: https://github.com/winlibs/libjpeg/blob/master/jerror.c#L69
exit() is not caught by gdb even if code is not 0 indeed._

File: src/Makefile.am [hunk 2]
```diff
@@ -191,6 +191,8 @@ BITCOIN_CORE_H = \
   governance/validators.h \
   governance/vote.h \
   governance/votedb.h \
+  gsl/assert.h \
+  gls/pointers.h \
```
Comment (UdjinM6) [inline src/Makefile.am:195]:
```suggestion
  gsl/pointers.h \
```

Comment (PastaPastaPasta) [reply src/Makefile.am:195]: whoops :D I hate that issues like this still compiles locally

File: src/llmq/quorums.cpp [hunk 3]
```diff
@@ -23,6 +23,7 @@
 #include <util/underlying.h>
 #include <validation.h>
 
+#include <gsl/pointers.h>
```
Comment (UdjinM6) [inline src/llmq/quorums.cpp:26]:
```suggestion
```
you already have one in `src/llmq/quorums.h`

File: src/source_location.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2023 The Dash Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+// Originally from https://github.com/paweldac/source_location
```
Comment (UdjinM6) [inline src/source_location.h:4]: That's basically the original file with no changes but in whitespaces and include protections, right? We should just include the original version here and add `source_location.h` to linter exceptions like we do for every other external library (and like you did for `gsl/` already) imo.

Comment (PastaPastaPasta) [reply src/source_location.h:4]: Yes, pretty much, but it didn't have an explicit MIT license in file; only in repo, so I thought the was good to add; also good to have a link to the source; and then I figured it'd be easy to just change the include guards than bother with the linter. IDK, do you really want any of that changed?

Comment (UdjinM6) [reply src/source_location.h:4]: hmm... makes sense I guess. ok, let's keep it as is.

File: src/llmq/instantsend.cpp [hunk 5]
```diff
@@ -615,13 +616,11 @@ bool CInstantSendManager::CheckCanLock(const COutPoint& outpoint, bool printDebu
...
-    int nTxAge;
-    {
+    auto [pindexMined, nTxAge] = [&]() {
         LOCK(cs_main);
-        pindexMined = m_chainstate.m_blockman.LookupBlockIndex(hashBlock);
-        nTxAge = m_chainstate.m_chain.Height() - pindexMined->nHeight + 1;
-    }
+        auto pindex = m_chainstate.m_blockman.LookupBlockIndex(hashBlock);
+        return std::make_pair(pindex, m_chainstate.m_chain.Height() - pindex->nHeight + 1);
+    }();
```
Comment (UdjinM6) [inline src/llmq/instantsend.cpp:623]: looks unrelated

Comment (PastaPastaPasta) [reply src/llmq/instantsend.cpp:623]: but is it bad? Yeah... It is pretty unrelated it seems.

**Comment by UdjinM6** [comment]
let's maybe have both cout and LogPrintf? https://github.com/UdjinM6/dash/commit/4eb326cb3d0c03b0d8dd114ac9014792850f3209

**Comment by knst** [comment]
It should not be stdout, but stderr

**Review by UdjinM6 — APPROVED** [review]
ACK

File: src/gsl/assert.h [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+#endif // defined(GSL_MSVC_USE_STL_NOEXCEPTION_WORKAROUND)
+
+        [[noreturn]] inline void terminate(nostd::source_location loc) noexcept
+        {
+#if defined(GSL_MSVC_USE_STL_NOEXCEPTION_WORKAROUND)
+            (*gsl::details::get_terminate_handler())();
+#else
+            std::ostringstream s;
+            s << "ERROR: error detected null not_null detected at " << loc.file_name() << ":" << loc.line() << ":"
+                                                                    << loc.column() << ":" << loc.function_name() << "\n";
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/gsl/assert.h]: Yeah, we could do 3d69147555. I doubt it would make any difference but 🤷‍♂️

**Review by UdjinM6 — APPROVED** [review]
re-ACK

**Comment by PastaPastaPasta** [comment]
LGTM
